openraft/tests/life_cycle/t20_initialization.rs

```diff
@@ -235,6 +235,74 @@ async fn initialize_err_not_allowed() -> anyhow::Result<()> {
     Ok(())
 }
 
+/// What does this test do?
+///   this test is ought to test if the router is aware of network failures
+///
+/// - initialize 1 node as leader and 2 nodes as follower
+/// - set one follower's network to be unreachable
+/// - check if router's connecting to unreachable follower returning an `Err`
+/// - assert whether the cluster still works properly
+#[async_entry::test(worker_threads = 8, init = "init_default_ut_tracing()", tracing_span = "debug")]
+async fn router_network_failure_aware() -> anyhow::Result<()> {
+    let config = Arc::new(Config::default().validate()?);
+    let nodes = btreeset! {0, 1, 2};
+    let mut router = RaftRouter::new(config.clone());
+
+    let log_index = router.new_nodes_from_single(nodes.clone(), btreeset!()).await?;
+
+    tracing::info!("--- add unreachable learner to cluster");
+    {
+        router.new_raft_node(3);
+        router.enable_connect(3, false);
+        assert!(router.add_learner(0, 3).await.is_err());
+
+        router
+            .wait_for_log(
+                &btreeset! {0, 1, 2},
+                Some(log_index),
+                timeout(),
+                "add learner fail, no logs",
+            )
+            .await?;
+    }
+
+    tracing::info!("--- finish making cluster");
+    {
+        router
+            .wait_for_members(
+                &btreeset! {0},
+                nodes.clone(),
+                timeout(),
+                format!("cluster should be: {:?}", nodes.clone()).as_str(),
+            )
+            .await?;
+    }
+
+    tracing::info!("--- write 10 logs");
+    {
+        router.client_request_many(0, "client", 10).await?;
+    }
+
+    tracing::info!("--- block n2, make it unreachable for router");
+    {
+        router.enable_connect(2, false);
+    }
+
+    tracing::info!("--- isolate leader, force an election");
+    {
+        router.isolate_node(0);
+    }
+
+    tracing::info!("--- cluster should unable to commit logs now");
+    {
+        let write_req = router.client_request(1, "client", 0);
+        let timeout = tokio::time::timeout(Duration::from_millis(1000), write_req).await;
+        assert!(timeout.is_err(), "now cluster should fail");
+    }
```

Comment:
The test has to assert that a cluster should work with a minority of nodes that can not connect.